src/decycle.ts

```diff
@@ -1,36 +1,38 @@
 import { isPlainObject } from './isPlainObject';
 import { pathToPointer } from './pathToPointer';
 
-export const decycle = (obj: unknown, replacer?: (value: any) => any) => {
+export function decycle(obj: unknown, replacer?: (value: any) => any) {
   const objs = new WeakMap<object, string>();
-  return (function derez(value: any, path: string[]) {
-    // The new object or array
-    let curObj: any;
-
-    // If a replacer function was provided, then call it to get a replacement value.
-    if (replacer) value = replacer(value);
+  function derez(value: any, path: (string | number)[]): any {
+    if (replacer) {
+      value = replacer(value);
+    }
 
     if (isPlainObject(value) || Array.isArray(value)) {
-      // The path of an earlier occurance of value
       const oldPath = objs.get(value);
-      // If the value is an object or array, look to see if we have already
-      // encountered it. If so, return a {"$ref":PATH} object.
-      if (oldPath) return { $ref: oldPath };
+
+      if (oldPath) {
+        return { $ref: oldPath };
+      }
 
       objs.set(value, pathToPointer(path));
-      // If it is an array, replicate the array.
+
       if (Array.isArray(value)) {
-        curObj = value.map((element, i) => derez(element, [...path, String(i)]));
-      } else {
-        // It is an object, replicate the object.
-        curObj = {};
-        Object.keys(value).forEach(name => {
-          curObj[name] = derez(value[name], [...path, name]);
-        });
+        return value.map((element, i) => derez(element, [...path, i]));
```

Comment:
This code doesn't do exactly what the previous version did:
1. since you returning a value here this line `objs.delete(value);` will never be reached.
2. map callback fn casted index to string previously: `derez(element, [...path, String(i)]))`. It doesn't do this anymore.

Reply:
Its working as all test cases are passed and if we removed that test case will be failed

Reply:
are you sure that this feature is covered by the tests?

Reply:
Yes @rainum its already there, I just did some changes optimisation nothing else

Reply:
I'm still having some doubts about this part: `derez(element, [...path, String(i)]))`. But if you say that this is covered by the tests, I think we can live with it.